File: 1025-minimum-cost-for-tickets/1025-minimum-cost-for-tickets.cpp
class Solution {
public:
    unordered_set<int> travelNeeded;
    int solve(vector<int>& dp, vector<int>& costs, int& lastDay, int currDay){
        if(currDay > lastDay) return 0;
        if(travelNeeded.find(currDay) == travelNeeded.end()) return solve(dp, costs, lastDay, currDay+1);
        if(dp[currDay] != -1) return dp[currDay];
        
        int oneDay = costs[0] + solve(dp, costs, lastDay, currDay+1);
        int sevenDay = costs[1] + solve(dp, costs, lastDay, currDay+7);
        int thirtyDay = costs[2] + solve(dp, costs, lastDay, currDay+30);

        dp[currDay] = min(oneDay, min(sevenDay, thirtyDay));
        
        return dp[currDay];
    }
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int lastDay = days.back();
        vector<int> dp(lastDay+1,-1);
        for(auto day: days){
            travelNeeded.insert(day);
        }
        return solve(dp, costs, lastDay, 1);
    }
};

File: 1025-minimum-cost-for-tickets/1025-minimum-cost-for-tickets.py
class Solution(object):
    def __init__(self):
        self.travelNeeded = set()
    
    def solve(self, dp, costs, lastDay, currDay):
        if currDay > lastDay:
            return 0
        if currDay not in self.travelNeeded:
            return self.solve(dp, costs, lastDay, currDay+1)
        if dp[currDay] != -1:
            return dp[currDay]
        
        oneDay = costs[0] + self.solve(dp, costs, lastDay, currDay+1)
        sevenDay = costs[1] + self.solve(dp, costs, lastDay, currDay+7)
        thirtyDay = costs[2] + self.solve(dp, costs, lastDay, currDay+30)

        dp[currDay] = min(oneDay, min(sevenDay, thirtyDay))
        return dp[currDay]
    
    def mincostTickets(self, days, costs):
        """
        :type days: List[int]
        :type costs: List[int]
        :rtype: int
        """
        lastDay = days[-1]
        dp = [-1]* (lastDay+1)
        for i in days:
            self.travelNeeded.add(i)
        return self.solve(dp, costs, lastDay, 1)
        